include/counter.h

```diff
@@ -39,6 +39,31 @@ extern "C" {
 
 /**@} */
 
+/**@defgroup COUNTER_TOP_FLAGS Flags used by @ref counter_top_cfg.
+ * @{
+ */
+
+/**
+ * @brief Flag preventing counter reset when top value is changed.
+ *
+ * If flags is set then counter is free running while top value is updated
+ * otherwise counter is reset. If counter is counting up and free running it is
+ * possible that new top value is smaller than current counter value. In that
+ * case counter may go out of bounds. @ref counter_set_top_value() will return
+ * error -ETIME if such situation is detected.
+ */
+#define COUNTER_TOP_CFG_DONT_RESET BIT(0)
+
+/**
+ * @brief Flag instructing counter to reset itself if changing top value
+ *	  results in counter going out of new top value bound.
+ *
+ * See @ref COUNTER_TOP_CFG_DONT_RESET.
+ */
+#define COUNTER_TOP_CFG_RESET_WHEN_LATE BIT(1)
```

Comment:
There is one difference in the description and implementation of `COUNTER_TOP_CFG_RESET_WHEN_LATE` and an equivalent `COUNTER_ALARM_CFG_EXPIRE_WHEN_LATE` flag introduced in #16252. The former resets the counter and returns `-ETIME` error in case late setting is detected the latter doesn't. The behavior should probably be aligned.

Reply:
agree. I lean towards returning error then. User may want system to continue to work but at the same time have trace of this rather not expected situation.